<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>plotting – *bio*DSC</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../images/bioDSC-without-dots-transparent.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/bioDSC-without-dots-transparent.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title"><em>bio</em>DSC</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../workshops.html"> 
<span class="menu-text">Workshops</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../people.html"> 
<span class="menu-text">People</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../tutorials.html"> 
<span class="menu-text">Tutorials</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../further-reading.html"> 
<span class="menu-text">Further reading</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://www.uva.nl" target="_blank"> 
<span class="menu-text"><img src="../../images/uva-logo.png" height="30"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#exercises-lesson-9-plotting" id="toc-exercises-lesson-9-plotting" class="nav-link active" data-scroll-target="#exercises-lesson-9-plotting">Exercises Lesson 9: Plotting</a>
  <ul class="collapse">
  <li><a href="#minima-and-maxima" id="toc-minima-and-maxima" class="nav-link" data-scroll-target="#minima-and-maxima">Minima and maxima</a></li>
  <li><a href="#mean-gene-expression" id="toc-mean-gene-expression" class="nav-link" data-scroll-target="#mean-gene-expression">Mean gene expression</a></li>
  <li><a href="#more-correlations" id="toc-more-correlations" class="nav-link" data-scroll-target="#more-correlations">More correlations</a></li>
  </ul></li>
  <li><a href="#additional-exercises" id="toc-additional-exercises" class="nav-link" data-scroll-target="#additional-exercises">Additional Exercises</a>
  <ul class="collapse">
  <li><a href="#even-more-correlations" id="toc-even-more-correlations" class="nav-link" data-scroll-target="#even-more-correlations">Even more correlations</a></li>
  <li><a href="#subselection-and-melting" id="toc-subselection-and-melting" class="nav-link" data-scroll-target="#subselection-and-melting">Subselection and melting</a></li>
  <li><a href="#saving-your-plot" id="toc-saving-your-plot" class="nav-link" data-scroll-target="#saving-your-plot">Saving your plot</a></li>
  <li><a href="#correlations" id="toc-correlations" class="nav-link" data-scroll-target="#correlations">Correlations</a></li>
  <li><a href="#normalized-dataframe" id="toc-normalized-dataframe" class="nav-link" data-scroll-target="#normalized-dataframe">Normalized dataframe</a></li>
  <li><a href="#crude-oil" id="toc-crude-oil" class="nav-link" data-scroll-target="#crude-oil">Crude oil</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">




<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Credits &amp; copyright
</div>
</div>
<div class="callout-body-container callout-body">
<p>Many questions originate from the carpentry lesson <a href="http://swcarpentry.github.io/python-novice-gapminder/">“Plotting and Programming in Python”</a>, some were authored by the <em>bio</em>DSC. Carpentries website materials are <a href="https://software-carpentry.org/license/">licenced</a> under the <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>, which also applies to the contents of this website.</p>
</div>
</div>
<hr>
<!-- ####################################################################### -->
<section id="exercises-lesson-9-plotting" class="level1">
<h1>Exercises Lesson 9: Plotting</h1>
<section id="minima-and-maxima" class="level3">
<h3 class="anchored" data-anchor-id="minima-and-maxima">Minima and maxima</h3>
<div id="question-minmax" class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Fill in the blanks below to plot the minimum GDP per capita over time for all the countries in Europe.</p>
<p>Modify it again to plot the maximum GDP per capita over time for Europe, you need to edit the code beyond the <code>___</code> for this.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>data_europe <span class="op">=</span> pd.read_csv(<span class="st">'data/gapminder_gdp_europe.csv'</span>, index_col<span class="op">=</span><span class="st">'country'</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>data_europe_transposed <span class="op">=</span> data_europe.T</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>data_europe_transposed[<span class="st">'min'</span>] <span class="op">=</span> data_europe.____</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>data_europe_transposed[<span class="st">'max'</span>] <span class="op">=</span> ____</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>data_europe_transposed[<span class="st">'year'</span>] <span class="op">=</span> ____</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>sns.lineplot(data_europe_transposed, x<span class="op">=</span><span class="st">'year'</span>, y<span class="op">=</span><span class="st">'min'</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>sns.lineplot(data_europe_transposed, x<span class="op">=</span><span class="st">'year'</span>, y<span class="op">=</span><span class="st">'max'</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="st">'best'</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>HINT: if you don’t see the solution, take it step by step. Break down the task in subtasks, and adress the first step towards the solution first. Try that first. Running code is free.</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data_europe <span class="op">=</span> pd.read_csv(<span class="st">'data/gapminder_gdp_europe.csv'</span>, index_col<span class="op">=</span><span class="st">'country'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data_europe_transposed <span class="op">=</span> data_europe.T</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>data_europe_transposed[<span class="st">'min'</span>] <span class="op">=</span> data_europe.<span class="bu">min</span>()</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>data_europe_transposed[<span class="st">'max'</span>] <span class="op">=</span> data_europe.<span class="bu">max</span>()</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>data_europe_transposed[<span class="st">'year'</span>] <span class="op">=</span> data_europe_transposed.index.<span class="bu">str</span>.replace(<span class="st">"gdpPercap_"</span>,<span class="st">""</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>sns.lineplot(data_europe_transposed, x<span class="op">=</span><span class="st">'year'</span>, y<span class="op">=</span><span class="st">'min'</span>, label<span class="op">=</span><span class="st">'min'</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>sns.lineplot(data_europe_transposed, x<span class="op">=</span><span class="st">'year'</span>, y<span class="op">=</span><span class="st">'max'</span>, label<span class="op">=</span><span class="st">'max'</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="st">'best'</span>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The figure now looks like:</p>
<p><img src="../../static/plots_pyworkshop/L9_min_max_gdp_europe.png" class="img-fluid" style="width:40.0%"></p>
</div>
</div>
</div>
<!-- Carpentries credits L9, edited -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: <a href="https://swcarpentry.github.io/python-novice-gapminder/09-plotting.html" style="color: #777777" target="_blank">Carpentries workshop materials</a>, edited by the </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
</section>
<section id="mean-gene-expression" class="level3">
<h3 class="anchored" data-anchor-id="mean-gene-expression">Mean gene expression</h3>
<p>Use the kohela data again, and process it as such:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data, note the ".T" at the end here</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>df_kohela <span class="op">=</span> pd.read_csv(<span class="st">'data/kohela-et-al.csv'</span>, index_col<span class="op">=</span><span class="dv">0</span>).T</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create new 'masks'</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>epicardial_cells <span class="op">=</span> df_kohela[<span class="st">'WT1'</span>]<span class="op">&gt;</span><span class="dv">3</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>fibroblast_cells <span class="op">=</span> df_kohela[<span class="st">'COL2A1'</span>]<span class="op">&gt;</span><span class="dv">30</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>fat_cells <span class="op">=</span> df_kohela[<span class="st">'PPARG'</span>]<span class="op">&gt;</span><span class="dv">2</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Add cell type</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>df_kohela[<span class="st">'Celltype'</span>] <span class="op">=</span> <span class="st">'unknown'</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>df_kohela.loc[epicardial_cells,<span class="st">'Celltype'</span>] <span class="op">=</span> <span class="st">'epicardial'</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>df_kohela.loc[fibroblast_cells, <span class="st">'Celltype'</span>] <span class="op">=</span> <span class="st">'fibroblast'</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>df_kohela.loc[fat_cells, <span class="st">'Celltype'</span>] <span class="op">=</span> <span class="st">'fat'</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Add conditions</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>df_kohela[<span class="st">'Condition'</span>] <span class="op">=</span> <span class="st">'unknown'</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>df_kohela.loc[df_kohela.index.<span class="bu">str</span>.contains(<span class="st">'WT_'</span>), <span class="st">'Condition'</span>] <span class="op">=</span> <span class="st">'WT'</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>df_kohela.loc[df_kohela.index.<span class="bu">str</span>.contains(<span class="st">'mutant_'</span>), <span class="st">'Condition'</span>] <span class="op">=</span> <span class="st">'mutant'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>Disclaimer: we are analyzing single cell data in this exercise. Within the constraints of this introductory course, we don’t use proper statistic or methodological approaches to analyze the data. These exercises are meant to teach you Python concepts, but also give you a flavor of biological data analysis.</em></p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question A
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use seaborn to create a ‘stripplot’ plot for WT1 expression per cell type. Then create a similar plot for TBX18 (both epicardial cell markers). What information can be extracted from this plot?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>sns.stripplot(df_kohela, x<span class="op">=</span><span class="st">'Celltype'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>, jitter<span class="op">=</span><span class="va">True</span>, color<span class="op">=</span><span class="st">'red'</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, rotation<span class="op">=</span><span class="dv">45</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sns.stripplot(df_kohela, x<span class="op">=</span><span class="st">'Celltype'</span>, y<span class="op">=</span><span class="st">'WT1'</span>, jitter<span class="op">=</span><span class="va">True</span>, color<span class="op">=</span><span class="st">'blue'</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, rotation<span class="op">=</span><span class="dv">45</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="../../static/plots_pyworkshop/L9_stripplot_WT1_TBX18.png" class="img-fluid" style="width:50.0%"></p>
<p>Since we classified cells with high WT1 levels as epicardial cells, we expected to see high WT1 levels in epicardial cells by construction. This is indeed the case.</p>
<p>We see one stray fibroblast cell with a high WT1 level. This cell appears to be WT1+ <strong>and</strong> COL2A1+. Because of the order in which we executed the code, this cell got classified as fibroblast.</p>
<p>TBX18 is expected to show high expression in epicardial cells too, and, if both WT1 and TBX18 are specific for epicardial cells, they should not occur in other cell types. This also appears to be the case.</p>
<p>Quite some cells still have the ‘unknown’ phenotype, but otherwise these plots are supportive of our strategy of using WT1 and TBX18 as epicardial markers.</p>
<p>For a real experiment, you might use packages like <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0">scanpy</a> (Python) or <a href="https://satijalab.org/seurat/">Seurat</a> (R) to analyze your data. The authors of the <a href="https://doi.org/10.1126/scitranslmed.abf2750">Kohela et al.</a> study used <a href="https://cole-trapnell-lab.github.io/monocle-release/">Monocle 2</a> (R).</p>
<!-- The plots shows that the expression in the epicardial and unknown types is the largest and that comparatively the fat and fibroblasts have low expression levels. -->
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question B
</div>
</div>
<div class="callout-body-container callout-body">
<p>Now create a scatter plot, showing WT1 expression vs.&nbsp;TBX18 expression across all cells. What does this tell us?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Creating a scatterplot is straightforward:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(df_kohela, x<span class="op">=</span><span class="st">'WT1'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="../../static/plots_pyworkshop/L9_scatter_WT1_TBX18.png" class="img-fluid" style="width:25.0%"></p>
<p>This plot highlights a few challenges with data analysis.</p>
<p>Single cell RNA-seq data can typically show only a few reads per gene per cell, or even zero, even when the gene is expressed. This is a limitation of the RNA-seq technique. This results in many cells having only few or zero counts for the WT1 and TBX18 gene expression. In addition, the high number number of cells results in many overlapping datapoints. It is therefore hard to see any pattern in the data.</p>
<p>We can make the individual point slightly transparent by adding <code>alpha=0.1</code> as argument to the scatterplot method. This already helps to visualize the data better:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(df_kohela, x<span class="op">=</span><span class="st">'WT1'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>, alpha<span class="op">=</span><span class="fl">.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="../../static/plots_pyworkshop/L9_scatter_WT1_TBX18_alpha.png" class="img-fluid" style="width:25.0%"></p>
<p>It appears that there is not much correlation between the TBX18 and WT1 expression. This can either be because cells do not typically express both at the same time, or because our detection is not sufficient to identify such a correlation from this plot.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question C
</div>
</div>
<div class="callout-body-container callout-body">
<p>Color individual datapoints of the scatter plot per cell type. What does this tell us?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(df_kohela, x<span class="op">=</span><span class="st">'WT1'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>, hue<span class="op">=</span><span class="st">'Celltype'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="section" class="level1">
<h1><img src="../../static/plots_pyworkshop/L9_scatter_WT1_TBX18_Celltype.png" class="img-fluid" style="width:40.0%"></h1>
<p>This plot is still somewhat hard to interpret, because many points are overlaying on top of each other.</p>
<p>One way to disentangle the points, is to plot the different celltypes separately:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>maxval <span class="op">=</span> np.<span class="bu">max</span>(df_kohela.loc[:,[<span class="st">'WT1'</span>,<span class="st">'TBX18'</span>]])</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Epicardial cells'</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(df_kohela.loc[df_kohela[<span class="st">'Celltype'</span>]<span class="op">==</span><span class="st">'epicardial'</span>,:], x<span class="op">=</span><span class="st">'WT1'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>, alpha<span class="op">=</span><span class="fl">.1</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>plt.xlim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>plt.ylim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Fibroblast cells'</span>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(df_kohela.loc[df_kohela[<span class="st">'Celltype'</span>]<span class="op">==</span><span class="st">'fibroblast'</span>,:], x<span class="op">=</span><span class="st">'WT1'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>, alpha<span class="op">=</span><span class="fl">.1</span>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>plt.xlim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>plt.ylim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(df_kohela.loc[df_kohela[<span class="st">'Celltype'</span>]<span class="op">==</span><span class="st">'fat'</span>,:], x<span class="op">=</span><span class="st">'WT1'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>, alpha<span class="op">=</span><span class="fl">.1</span>)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Fat cells'</span>)</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>plt.xlim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>plt.ylim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(df_kohela.loc[df_kohela[<span class="st">'Celltype'</span>]<span class="op">==</span><span class="st">'unknown'</span>,:], x<span class="op">=</span><span class="st">'WT1'</span>, y<span class="op">=</span><span class="st">'TBX18'</span>, alpha<span class="op">=</span><span class="fl">.1</span>)</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Unknown cells'</span>)</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>plt.xlim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>plt.ylim([<span class="dv">0</span>, maxval<span class="op">+</span><span class="dv">1</span>])</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="../../static/plots_pyworkshop/L9_scatter_WT1_TBX18_Celltype-fibroblast.png" class="img-fluid" style="width:25.0%"> <img src="../../static/plots_pyworkshop/L9_scatter_WT1_TBX18_Celltype-epicardial.png" class="img-fluid" style="width:25.0%"></p>
<p><img src="../../static/plots_pyworkshop/L9_scatter_WT1_TBX18_Celltype-fat.png" class="img-fluid" style="width:25.0%"> <img src="../../static/plots_pyworkshop/L9_scatter_WT1_TBX18_Celltype-unknown.png" class="img-fluid" style="width:25.0%"></p>
<p>Interestingly, we now see that fibroblast (COL2A1+) and fat cells (PPARG+) typically don’t express WT1 or TBX18, whilst epicardial cells (WT1+) do also often show TBX18 expression.</p>
<p>There are nevertheless still WT+ cells that don’t show TBX18 expression, and moreover many WT- cells (the unknown cells) that show TBX18 expression, so these plots are not conclusive, and more sophisticated single cell methods might be needed to further study these cells.</p>
</section>
</div>
</div>
</div>
<!--
::: {.callout-tip icon='false'}
## Question D
1. Calculate the total RNA-seq reads per cell, and make a violin plot per condition for these total reads.

Some code to start with:
```python
df_kohela['Total_reads'] = df_kohela.loc[:,'A1BG':'ZZZ3']._.____
```

2. What does the violin plot tell us?

:::

::: {.callout-caution collapse='true' icon='false'}
## Answer

1. ```python
df_kohela['Total_reads'] = df_kohela.loc[:,'A1BG':'ZZZ3'].sum(axis=1)
```
2. ```python
sns.violinplot(df_kohela,x='Celltype',y='Total_reads')
```
This plot tells us that for each cell type the average total number of reads are comparable. However, there are some fibrobblast cells that show a very large read-count. Perhaps remove them or look closer which genes are responsible for this?
:::
-->
<!-- bioDSC credits -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
</section>
<section id="more-correlations" class="level3">
<h3 class="anchored" data-anchor-id="more-correlations">More correlations</h3>
<p>This short program creates a plot showing the correlation between GDP and life expectancy for 2007, scaling marker size by population:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>data_all <span class="op">=</span> pd.read_csv(<span class="st">'data/gapminder_all.csv'</span>, index_col<span class="op">=</span><span class="st">'country'</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data_all, x<span class="op">=</span><span class="st">'gdpPercap_2007'</span>, y<span class="op">=</span><span class="st">'lifeExp_2007'</span>, size<span class="op">=</span><span class="st">'pop_2007'</span>, sizes<span class="op">=</span>(<span class="dv">1</span>,<span class="dv">40</span><span class="op">**</span><span class="dv">2</span>), legend<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using online help and other resources, explain what each plotting argument does.</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>See e.g.&nbsp;<a href="https://seaborn.pydata.org/generated/seaborn.scatterplot.html">the scatterplot documentation</a>.</p>
<ul>
<li><code>data_all</code>: The first argument specifies which dataframe to use as input.</li>
<li><code>x='gdpPercap_2007'</code>: The x argument expects the name of the column that holds values to plot on the x-axis.</li>
<li><code>y='lifeExp_2007'</code>: Same as x, but for the y axis.</li>
<li><code>size='pop_2007'</code>: This tells seaborn to scale the point size with the values in a column, in this case the population sizes of the countries (in 2007).</li>
<li><code>sizes=(1,40**2)</code>: This sets the range of the sizes to be used for the points, given as (min, max). These are areas, which is why it’s convenient to use the square operator <code>**2</code>, such that you can choose a radius (in this case minimally 1 and maximally 40).</li>
<li><code>legend=False</code>: This turns off the legend.</li>
</ul>
<p>The result:</p>
<p><img src="../../static/plots_pyworkshop/L9_scatter_gdp_lifeexp.png" class="img-fluid" style="width:35.0%"></p>
</div>
</div>
</div>
<!-- Carpentries credits L9, edited -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: <a href="https://swcarpentry.github.io/python-novice-gapminder/09-plotting.html" style="color: #777777" target="_blank">Carpentries workshop materials</a>, edited by the </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
<!-- ####################################################################### -->
<hr>
</section>
</section>
<section id="additional-exercises" class="level1">
<h1>Additional Exercises</h1>
<!-- ####################################################################### -->
<section id="even-more-correlations" class="level3">
<h3 class="anchored" data-anchor-id="even-more-correlations">Even more correlations</h3>
<!-- bioDSC -->
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question A
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the code from the “More Correlations” exercise, and try to add the following lines to the plotting code:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>plt.text(data_all.loc[<span class="st">'United States'</span>,<span class="st">'gdpPercap_2007'</span>], data_all.loc[<span class="st">'United States'</span>,<span class="st">'lifeExp_2007'</span>], <span class="st">'United States'</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>plt.text(data_all.loc[<span class="st">'Netherlands'</span>,<span class="st">'gdpPercap_2007'</span>], data_all.loc[<span class="st">'Netherlands'</span>,<span class="st">'lifeExp_2007'</span>], <span class="st">'Netherlands'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>What’s happening here? (You might need to use Google.)</li>
<li>Add your favorite country too.</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The function <a href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.text.html">matplotlib.pyplot.text</a>,</p>
<pre><code>matplotlib.pyplot.text(x, y, s, fontdict=None, **kwargs)</code></pre>
<p>allows you to put text s in your plotting area at location x,y. x, y and s should be single values (not arrays or tables)</p>
<ul>
<li><code>data_all.loc['United States','gdpPercap_2007']</code> selects the x value corresponding to the united states.</li>
<li><code>data_all.loc['United States','lifeExp_2007']</code> selects the y value corresponding to the united states.</li>
<li><code>'United states'</code> provides the label.</li>
</ul>
<p>The result:</p>
<p><img src="../../static/plots_pyworkshop/L9_scatter_gdp_lifeexp_label.png" class="img-fluid" style="width:40.0%"></p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question B
</div>
</div>
<div class="callout-body-container callout-body">
<p>How would you add labels for the top 10 GDP countries?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>This would be very tedious with what you learned currently! In the next lessons (particularly lesson 12), we’ll learn how to automate your code. This will be very useful for this particular challenge.</p>
</div>
</div>
</div>
<!-- bioDSC credits -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
</section>
<section id="subselection-and-melting" class="level3">
<h3 class="anchored" data-anchor-id="subselection-and-melting">Subselection and melting</h3>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question A
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using the kohela data, first create a new dataframe <code>df_kohela_sel</code> with only a selection of a few genes, and the condition and cell type columns.</p>
<p>Select the following genes: <code>['WT1', 'TBX18', 'TFAP2A', 'COL2A1', 'ACTA2', 'PPARG', 'CEBPA']</code>. These are epicardial markers (WT1, TBX18), a transcription factor (TFAP2A), fibroblast markers (COL2A1, ACTA2), and fat markers (PPARG, CEBPA).</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>df_kohela_sel <span class="op">=</span> <span class="op">\</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    df_kohela.loc[:,[<span class="st">'WT1'</span>, <span class="st">'TBX18'</span>, <span class="st">'TFAP2A'</span>, <span class="st">'COL2A1'</span>, <span class="st">'ACTA2'</span>, <span class="st">'PPARG'</span>, <span class="st">'CEBPA'</span>,<span class="st">'Celltype'</span>,<span class="st">'Condition'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question B
</div>
</div>
<div class="callout-body-container callout-body">
<p>Now melt this dataframe using <code>pd.melt()</code>, and use cell type and condition as identifier variables.</p>
<ul>
<li>What will happen to the gene expression values?</li>
<li>What is sensible input for the <code>var_name</code> and <code>value_name</code> parameters?</li>
<li>Why is this useful? (For answer, see next questions.)</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>df_kohela_melted <span class="op">=</span> pd.melt(df_kohela_sel, <span class="op">\</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    id_vars<span class="op">=</span>[<span class="st">'Celltype'</span>,<span class="st">'Condition'</span>], <span class="op">\</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    var_name<span class="op">=</span><span class="st">'Gene'</span>, value_name<span class="op">=</span><span class="st">'Expression'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><em>What will happen to the gene expression values?</em>
<ul>
<li>Expression data that were spread out over multiple columns are now re-organized in one long column. A new column is created alongside the column with all data, that specifies from which column the data originated.</li>
</ul></li>
<li><em>What is sensible input for the var_name and value_name parameters?</em>
<ul>
<li>The <code>var_name</code> is the name of the new column that specifies from which original columns the data came. In this case the original columns corresponded to different genes, and so “Gene” might be a sensible name.</li>
<li><code>value_name</code> is the name of the new column that holds all values previously spread out over multiple columns. The values all related to expression, and so “Expression” seems a good name.</li>
</ul></li>
<li><em>Why is this useful? (For answer, see next questions.)</em>
<ul>
<li>This is useful because now this dataframe can be easily used as input for the seaborn plots.</li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question C
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use <code>df_kohela_melted.head()</code> to check whether the output is as expected.</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The output was:</p>
<pre><code>  Celltype Condition Gene  Expression
0  unknown    mutant  WT1           0
1  unknown    mutant  WT1           0
2  unknown    mutant  WT1           0
3  unknown    mutant  WT1           0
4  unknown    mutant  WT1           1</code></pre>
<p>and indeed, the columns Celltype and Condition were retained, and the expression data was re-organized into one column “Expression”, annotated by the new column “Gene” which lists from which original column that data originated.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question D
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make a violin or stripplot, with as x-axis the genes, expression on the y-axis, and colored for condition. Is there an issue with this plot?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The code:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sns.violinplot(df_kohela_melted, x<span class="op">=</span><span class="st">'Gene'</span>, y<span class="op">=</span><span class="st">'Expression'</span>, hue<span class="op">=</span><span class="st">'Condition'</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>sns.stripplot(df_kohela_melted, x<span class="op">=</span><span class="st">'Gene'</span>, y<span class="op">=</span><span class="st">'Expression'</span>, hue<span class="op">=</span><span class="st">'Condition'</span>, ax<span class="op">=</span>ax, dodge<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="for-the-violinplot" class="level6">
<h6 class="anchored" data-anchor-id="for-the-violinplot"><em>For the Violinplot</em></h6>
<p><img src="../../static/plots_pyworkshop/L9_violinplot_Gene_Expression_Condition.png" class="img-fluid" style="width:40.0%"></p>
<p>An issue you might have identified is that it is very hard to compare gene expression levels because some of these genes have rather high expression levels, thus forcing the expression levels of other genes to the very bottom of the plot, making it hard to distinguish differences.</p>
<p>Also some outliers appear to be affecting the scale to a large extend, making it hard to see trends.</p>
<p>Moreover, you can probably not even see the Violins..</p>
</section>
<section id="for-the-stripplot" class="level6">
<h6 class="anchored" data-anchor-id="for-the-stripplot"><em>For the stripplot</em></h6>
<p><img src="../../static/plots_pyworkshop/L9_stripplot_Gene_Expression_Condition.png" class="img-fluid" style="width:40.0%"></p>
<p>This looks better, but it is still hard to see differences between wild type and mutant cell gene expression.</p>
</section>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question E
</div>
</div>
<div class="callout-body-container callout-body">
<p>Look at the following example:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># A custom function, which normalizes a series by its mean</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We'll learn more about functions in Lesson 16</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> gene_normalization(X):</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> X <span class="op">/</span> np.mean(X)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a subset of the data</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>cell_subset <span class="op">=</span> [<span class="st">'mutant_rep1_cell174'</span>, <span class="st">'WT_rep2_cell348'</span>, <span class="st">'mutant_rep1_cell160'</span>,</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>       <span class="st">'WT_rep1_cell022'</span>, <span class="st">'mutant_rep1_cell069'</span>]</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>gene_subset <span class="op">=</span> [<span class="st">'WT1'</span>, <span class="st">'TBX18'</span>, <span class="st">'TFAP2A'</span>, <span class="st">'COL2A1'</span>, <span class="st">'ACTA2'</span>, <span class="st">'PPARG'</span>, <span class="st">'CEBPA'</span>]</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize gene expression</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>df_kohela_subset2 <span class="op">=</span> df_kohela.loc[cell_subset, gene_subset]</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>df_kohela_subset2_normalized <span class="op">=</span> df_kohela_subset2.transform(gene_normalization)</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the result</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df_kohela_subset2_normalized)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol type="1">
<li>Check out what <code>gene_normalization(df_kohela_subset2['WT1'])</code> does.</li>
<li>What does the transform method do in the above code?</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="check-out-what-gene_normalizationdf_kohela_subset2wt1-does." class="level6">
<h6 class="anchored" data-anchor-id="check-out-what-gene_normalizationdf_kohela_subset2wt1-does."><em>1. Check out what <code>gene_normalization(df_kohela_subset2['WT1'])</code> does.</em></h6>
<p>What <code>gene_normalization()</code> is described in the comment above the function, it normalizes a series by its mean. <code>df_kohela_subset2['WT1']</code> is the series of expression values related to WT1.</p>
<p><code>gene_normalization(df_kohela_subset2['WT1'])</code> will thus return WT1 expression values normalized by their mean.</p>
</section>
<section id="what-does-the-transform-method-do-in-the-above-code" class="level6">
<h6 class="anchored" data-anchor-id="what-does-the-transform-method-do-in-the-above-code"><em>2. What does the transform method do in the above code?</em></h6>
<p>Transform applies a function to each of the columns in a dataframe. So in this case it will go over each column, which correspond to <code>df_kohela_subset2['WT1']</code>, <code>df_kohela_subset2['TBX18']</code>, <code>df_kohela_subset2['TFAP2A']</code> and so forth, and feed that column to the custom function, which in this case will return values that are normalized. Those normalized values are then put back in the column the function was applied to.</p>
<p>Thus, the each column will now correspond to mean-divided expression (ie will be normalized).</p>
</section>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question F
</div>
</div>
<div class="callout-body-container callout-body">
<p>Edit the following code (replacing blanks by code) to normalize the gene expression by the total expression of each gene. Hint: look at exercise E.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>df_kohela_grouped <span class="op">=</span> df_kohela_melted.groupby(_______)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>df_kohela_melted[<span class="st">'Expression_normalized'</span>] <span class="op">=</span> df_kohela_grouped[<span class="st">'Expression'</span>]._______(gene_normalization)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then, similar to D, plot the normalized gene expression using both the <code>sns.barplot</code> and <code>sns.stripplot</code>. For the stripplot, use the additional parameter <code>dodge=True</code>.</p>
<p>The barplot looks nice, but does it contain all information?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-27-contents" aria-controls="callout-27" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-27" class="callout-27-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The normalization:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>df_kohela_grouped <span class="op">=</span> df_kohela_melted.groupby(<span class="st">'Gene'</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>df_kohela_melted[<span class="st">'Expression_normalized'</span>] <span class="op">=</span> df_kohela_grouped[<span class="st">'Expression'</span>].transform(gene_normalization)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The bar code hides the information about the distribution of the single data points. So perhaps it is best to combine those two plots:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sns.barplot(df_kohela_melted, x<span class="op">=</span><span class="st">'Gene'</span>, y<span class="op">=</span><span class="st">'Expression_normalized'</span>, </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>            hue<span class="op">=</span><span class="st">'Condition'</span>, ci<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>sns.stripplot(df_kohela_melted, x<span class="op">=</span><span class="st">'Gene'</span>, y<span class="op">=</span><span class="st">'Expression_normalized'</span>, </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>              hue<span class="op">=</span><span class="st">'Condition'</span>, dodge<span class="op">=</span><span class="va">True</span>, color<span class="op">=</span><span class="st">'black'</span>, alpha<span class="op">=</span><span class="fl">.1</span>, size<span class="op">=</span><span class="dv">4</span>, </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>              legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>plt.ylim([<span class="dv">0</span>,<span class="dv">15</span>])</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>plt.close(<span class="st">'all'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The plot now looks like:</p>
<p><img src="../../static/plots_pyworkshop/L9_bar_stripplot_Gene_Expression_Normalized_Condition.png" class="img-fluid" style="width:40.0%"></p>
<p>Note that some data points are now hidden due to the choice of y limits. For a paper, one might need to think more deeply about how to display this data.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question G
</div>
</div>
<div class="callout-body-container callout-body">
<p>Finally, you might want to change the order in these plots. Use the following line:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>df_kohela_melted[<span class="st">'Condition'</span>] <span class="op">=</span> pd.Categorical(df_kohela_melted[<span class="st">'Condition'</span>], categories<span class="op">=</span>[<span class="st">'WT'</span>, <span class="st">'mutant'</span>], ordered<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>And then run your plotting code again. What has happened?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-29-contents" aria-controls="callout-29" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-29" class="callout-29-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>This results in the condition becoming a categorical series, which has a specific order, that is translated to the plot. This is very useful if you want to control where the labels go on your axes. The order of ‘WT’ and ‘mutant’ on the x-axis now makes more sense.</p>
<p>The figure now looks like:</p>
<p><img src="../../static/plots_pyworkshop/L9_bar_stripplot_Gene_Expression_Normalized_Condition_order.png" class="img-fluid" style="width:40.0%"></p>
</div>
</div>
</div>
<!-- bioDSC credits -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
</section>
<section id="saving-your-plot" class="level3">
<h3 class="anchored" data-anchor-id="saving-your-plot">Saving your plot</h3>
<p>You might want to save your plot. You can use the <code>plt.savefig</code> function for this.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Check out this code with some additional convenient options. Change <code>'/your/location/your-filename.pdf'</code> to a convenient path where you save your figure.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Bang Wong colorblind-friendly color scheme (https://www.nature.com/articles/nmeth.1618)</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>colors_bangwong <span class="op">=</span> [</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#E69F00"</span>,  <span class="co"># Orange</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#56B4E9"</span>,  <span class="co"># Sky Blue</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#009E73"</span>,  <span class="co"># Bluish Green</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#F0E442"</span>,  <span class="co"># Yellow</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#0072B2"</span>,  <span class="co"># Blue</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#D55E00"</span>,  <span class="co"># Vermillion</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#CC79A7"</span>,  <span class="co"># Reddish Purple</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#000000"</span>   <span class="co"># Black</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>plt.style.use(<span class="st">'default'</span>)</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">10</span><span class="op">/</span><span class="fl">2.54</span>,<span class="dv">10</span><span class="op">/</span><span class="fl">2.54</span>))</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>ax.plot([<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>], [<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">9</span>,<span class="dv">16</span>], linestyle<span class="op">=</span><span class="st">'--'</span>, color<span class="op">=</span>colors_bangwong[<span class="dv">1</span>], label<span class="op">=</span><span class="vs">r'$x^2$'</span>)</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>ax.plot([<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>], [<span class="dv">1</span>,<span class="dv">5</span>,<span class="dv">11</span>,<span class="dv">19</span>], linestyle<span class="op">=</span><span class="st">':'</span>, color<span class="op">=</span>colors_bangwong[<span class="dv">2</span>], label<span class="op">=</span><span class="vs">r'$x^2+(x-1)$'</span>)</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'X-axis'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Y-axis'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="st">'Sample Plot'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>ax.legend(fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>ax.tick_params(axis<span class="op">=</span><span class="st">'both'</span>, which<span class="op">=</span><span class="st">'major'</span>, labelsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'/your/location/your-filename.pdf'</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a>plt.close(fig)</span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Use this command to show the figure when not using (Jupyter) notebooks.</span></span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.show()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<!-- bioDSC credits -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
</section>
<section id="correlations" class="level3">
<h3 class="anchored" data-anchor-id="correlations">Correlations</h3>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question A
</div>
</div>
<div class="callout-body-container callout-body">
<p>Modify the code from “Minima and Maxima” exercise to create a scatter plot showing the relationship between the minimum and maximum GDP per capita across the countries in Asia, with each point in the plot corresponding to a year. What relationship do you see (if any)?</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-32-contents" aria-controls="callout-32" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-32" class="callout-32-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The code:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>data_asia <span class="op">=</span> pd.read_csv(<span class="st">'data/gapminder_gdp_asia.csv'</span>, index_col<span class="op">=</span><span class="st">'country'</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>data_asia_transposed <span class="op">=</span> data_asia.T</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>data_asia_transposed[<span class="st">'min'</span>] <span class="op">=</span> data_asia.<span class="bu">min</span>()</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>data_asia_transposed[<span class="st">'max'</span>] <span class="op">=</span> data_asia.<span class="bu">max</span>()</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>data_asia_transposed[<span class="st">'year'</span>] <span class="op">=</span> data_asia_transposed.index.<span class="bu">str</span>.replace(<span class="st">"gdpPercap_"</span>,<span class="st">""</span>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data_asia_transposed, x<span class="op">=</span><span class="st">'min'</span>, y<span class="op">=</span><span class="st">'max'</span>, hue<span class="op">=</span><span class="st">'year'</span>, palette<span class="op">=</span><span class="st">'viridis'</span>, legend<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Results in the plot:</p>
<p><img src="../../static/plots_pyworkshop/L9_scatter_min_max_gdp_asia.png" class="img-fluid" style="width:25.0%"></p>
<p>It can be seen that there is no correlation between the minimum GDP and maximum GDP for a specific year, indicating that GDPs across asia do not tend to rise and fall together.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question B
</div>
</div>
<div class="callout-body-container callout-body">
<p>You might note that the variability in the maximum is much higher than that of the minimum. Take a look at the maximum over time and the max indexes:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>data_asia <span class="op">=</span> pd.read_csv(<span class="st">'data/gapminder_gdp_asia.csv'</span>, index_col<span class="op">=</span><span class="st">'country'</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>df_max_GDP <span class="op">=</span> pd.DataFrame()</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>df_max_GDP[<span class="st">'GDP_max'</span>] <span class="op">=</span> data_asia.<span class="bu">max</span>()</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>df_max_GDP[<span class="st">'Year'</span>]    <span class="op">=</span> data_asia.columns.<span class="bu">str</span>.replace(<span class="st">'gdpPercap_'</span>,<span class="st">''</span>).astype(<span class="bu">int</span>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>plt.plot(df_max_GDP[<span class="st">'Year'</span>], df_max_GDP[<span class="st">'GDP_max'</span>])</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(data_asia.idxmax())</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(data_asia.idxmin())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-34-contents" aria-controls="callout-34" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-34" class="callout-34-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The plot looks like:</p>
<p><img src="../../static/plots_pyworkshop/L9_plot_max_GDP_asia.png" class="img-fluid" style="width:40.0%"></p>
<p>Seems the variability in this value is due to a sharp drop after 1972. Some geopolitics at play perhaps? Given the dominance of oil producing countries, maybe the Brent crude index would make an interesting comparison? Whilst Myanmar consistently has the lowest GDP, the highest GDP nation has varied more notably.</p>
</div>
</div>
</div>
<!-- Carpentries credits L9, edited -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: <a href="https://swcarpentry.github.io/python-novice-gapminder/09-plotting.html" style="color: #777777" target="_blank">Carpentries workshop materials</a>, edited by the </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
</section>
<section id="normalized-dataframe" class="level3">
<h3 class="anchored" data-anchor-id="normalized-dataframe">Normalized dataframe</h3>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>In the previous lesson about dataframes (in the additional exercises), we normalized the GDP data against the average trend. Plot the data from this normalized dataframe.</li>
<li>Is this helpful in any way?</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-36-contents" aria-controls="callout-36" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-36" class="callout-36-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the df to modify it and not interfer with other code</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>data_europe_relative_copy <span class="op">=</span> data_europe_relative.copy()</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the country as column</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>data_europe_relative_copy[<span class="st">'Country'</span>] <span class="op">=</span> data_europe_relative_copy.index</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Melt it </span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>data_europe_relative_melted <span class="op">=</span> data_europe_relative_copy.melt(id_vars<span class="op">=</span><span class="st">'Country'</span>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the year as number</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>data_europe_relative_melted[<span class="st">'year'</span>] <span class="op">=</span> data_europe_relative_melted[<span class="st">'variable'</span>].<span class="bu">str</span>.replace(<span class="st">'gdpPercap_'</span>,<span class="st">''</span>).astype(<span class="bu">int</span>)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="co"># And plot</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>sns.lineplot(data_europe_relative_melted, x<span class="op">=</span><span class="st">'year'</span>, y<span class="op">=</span><span class="st">'value'</span>, hue<span class="op">=</span><span class="st">'Country'</span>, legend<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We can now visualize the relative position of countries over time. Unfortunately, there are a little bit too many countries to clearly identify single ones. A solution to this issue is left for another time.</p>
<p><img src="../../static/plots_pyworkshop/L9_lineplot_europe_relative.png" class="img-fluid" style="width:40.0%"></p>
</div>
</div>
</div>
<!-- bioDSC credits -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>
<!-- ####################################################################### -->
</section>
<section id="crude-oil" class="level3">
<h3 class="anchored" data-anchor-id="crude-oil">Crude oil</h3>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Crude oil prices can be found <a href="https://www.eia.gov/dnav/pet/hist/LeafHandler.ashx?n=PET&amp;s=F000000__3&amp;f=A">here</a>.</p>
<p>This gives the data below:</p>
<pre><code>Decade  Year-0  Year-1  Year-2  Year-3  Year-4  Year-5  Year-6  Year-7  Year-8  Year-9
  1850's                                        16.00
  1860's    9.59    0.49    1.05    3.15    8.06    6.59    3.74    2.41    3.62    5.64
  1870's    3.86    4.34    3.64    1.83    1.17    1.35    2.52    2.38    1.17    0.86
  1880's    0.94    0.92    0.78    1.10    0.85    0.88    0.71    0.67    0.65    0.77
  1890's    0.77    0.56    0.51    0.60    0.72    1.09    0.96    0.68    0.80    1.13
  1900's    1.19    0.96    0.80    0.94    0.86    0.62    0.73    0.72    0.72    0.70
  1910's    0.61    0.61    0.74    0.95    0.81    0.64    1.10    1.56    1.98    2.01
  1920's    3.07    1.73    1.61    1.34    1.43    1.68    1.88    1.30    1.17    1.27
  1930's    1.19    0.65    0.87    0.67    1.00    0.97    1.09    1.18    1.13    1.02
  1940's    1.02    1.14    1.19    1.20    1.21    1.22    1.41    1.93    2.60    2.54
  1950's    2.51    2.53    2.53    2.68    2.78    2.77    2.79    3.09    3.01    2.90
  1960's    2.88    2.89    2.90    2.89    2.88    2.86    2.88    2.92    2.94    3.09
  1970's    3.18    3.39    3.39    3.89    6.87    7.67    8.19    8.57    9.00    12.64
  1980's    21.59   31.77   28.52   26.19   25.88   24.09   12.51   15.40   12.58   15.86
  1990's    20.03   16.54   15.99   14.25   13.19   14.62   18.46   17.23   10.87   15.56
  2000's    26.72   21.84   22.51   27.56   36.77   50.28   59.69   66.52   94.04   56.35
  2010's    74.71   95.73   94.52   95.99   87.39   44.39   38.29   48.05   61.40   55.59
  2020's    36.86   65.84   93.97   76.10                       </code></pre>
<p>Save that data to a .tsv file, and upload it.</p>
<p>Now try to understand the code below:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>df_crudeoil <span class="op">=</span> <span class="op">\</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    pd.read_csv(<span class="st">'data/crude_oil.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co"># reshape the data, such that it becomes a long list</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>df_crudeoil_melted <span class="op">=</span> df_crudeoil.melt(id_vars<span class="op">=</span><span class="st">'Decade'</span>, var_name<span class="op">=</span><span class="st">'lastdigit'</span>)</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="co"># now reformat the year information</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="co"># search and replace first</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>df_crudeoil_melted.loc[:,<span class="st">'Decade'</span>] <span class="op">=</span> df_crudeoil_melted.loc[:,<span class="st">'Decade'</span>].<span class="bu">str</span>.replace(<span class="st">"0's"</span>,<span class="st">''</span>)</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>df_crudeoil_melted.loc[:,<span class="st">'lastdigit'</span>] <span class="op">=</span> df_crudeoil_melted.loc[:,<span class="st">'lastdigit'</span>].<span class="bu">str</span>.replace(<span class="st">'Year-'</span>,<span class="st">''</span>)</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a><span class="co"># now combine information to create a new column "Year"</span></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>df_crudeoil_melted.loc[:,<span class="st">'Year'</span>] <span class="op">=</span> (df_crudeoil_melted.loc[:,<span class="st">'Decade'</span>] <span class="op">+</span> df_crudeoil_melted.loc[:,<span class="st">'lastdigit'</span>]).astype(<span class="bu">int</span>)</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the result</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df_crudeoil_melted.head())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using your new plotting skills, compare this data against the trends in the Asian GDPs showed earlier.</p>
</div>
</div>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-39-contents" aria-controls="callout-39" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-39" class="callout-39-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We saved to a <a href="../../static/data/crude_oil.tsv">tsv file</a> and ran the code above.</p>
<p>Then, made a plot:</p>
<pre><code>sns.lineplot(df_crudeoil_melted, x='Year', y='value', label='price')
plt.axvline(x=1972, color='red', linestyle='--', label='1972')
plt.legend()
plt.show()</code></pre>
<p>The plot looks like:</p>
<p><img src="../../static/plots_pyworkshop/L9_lineplot_crudeoil.png" class="img-fluid" style="width:40.0%"></p>
<p>Probably <a href="https://en.wikipedia.org/wiki/1973_oil_crisis">quite some things happened</a> to Arabian oil countries after ’72, related to oil and aformentioned drop in GDP.</p>
</div>
</div>
</div>
<!-- bioDSC credits -->
<div class="text-end">
<p><em><font color="#aaaaaa" size="1">Source: </font></em><font color="#aaaaaa" size="1">bio<em>DSC.</em></font></p>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>
<img src="../../images/uva-logo.png" height="25px;"> an <i>initiative</i> of the <a href="https://uva.nl" target="_blank">University of Amsterdam </a>
</p>
</div>
  </div>
</footer>




</body></html>